packages/repository/src/relations/has-many/has-many-through.repository.ts

```diff
@@ -141,36 +144,105 @@ export class DefaultHasManyThroughRepository<
+    const throughRepository = await this.getThroughRepository();
+    const throughConstraint = this.getThroughConstraint();
+    const throughInstances = await throughRepository.find(
+      constrainFilter(undefined, throughConstraint),
+      options?.throughOptions,
+    );
+    const targetConstraint = this.getTargetConstraint(throughInstances);
+    return targetRepository.find(
+      constrainFilter(filter, targetConstraint),
+      options,
+    );
   }
 
+  // 1. Given through links: {sourceId: 1, targetId: 1} and {sourceId: 2, targetId: 2}.
+  // When delete all models belonging to source id 1,
+  // then the target model id 1 and the first link is deleted.
+
+  // 2. Given trough links {sourceId: 1, targetId: 1}, {sourceId: 2, targetId: 1}
+  // and {sourceId: 2, targetId: 2}.
+  // When delete all models belonging to source id 1,
+  // then the target model id 1 and the first two links are deleted.
+  // Only {sourceId: 2, targetId: 2} link is preserved
   async delete(
     where?: Where<TargetEntity>,
     options?: Options & {
       throughOptions?: Options;
     },
   ): Promise<Count> {
-    throw new Error('Method not implemented.');
-  }
+    const targetRepository = await this.getTargetRepository();
+    const throughRepository = await this.getThroughRepository();
+    const throughConstraint = this.getThroughConstraint();
+    const throughInstances = await throughRepository.find(
+      constrainFilter(undefined, throughConstraint),
+      options?.throughOptions,
+    );
+    const targetConstraint = this.getTargetConstraint(throughInstances);
 
+    // delete throughs that have the targets that are going to be deleted
+    const throughFkConstraint = this.getThroughFkConstraint(
+      targetConstraint as TargetEntity,
+    );
+    await throughRepository.deleteAll(
+      constrainWhereOr({}, [throughConstraint, throughFkConstraint]),
+    );
+
+    // delete target(s)
+    return await targetRepository.deleteAll(
+      constrainWhere(where, targetConstraint as Where<TargetEntity>),
+      options,
+    );
+  }
+  // only allows patch target instances for now
   async patch(
     dataObject: DataObject<TargetEntity>,
     where?: Where<TargetEntity>,
     options?: Options & {
       throughOptions?: Options;
     },
   ): Promise<Count> {
-    throw new Error('Method not implemented.');
+    const targetRepository = await this.getTargetRepository();
+    const throughRepository = await this.getThroughRepository();
+    const throughConstraint = this.getThroughConstraint();
+    const throughInstances = await throughRepository.find(
+      constrainFilter(undefined, throughConstraint),
+      options?.throughOptions,
+    );
+    const targetConstraint = this.getTargetConstraint(throughInstances);
+    return targetRepository.updateAll(
+      constrainDataObject(dataObject, targetConstraint),
+      constrainWhere(where, targetConstraint as Where<TargetEntity>),
+      options,
+    );
   }
 
   async link(
```

Comment:
I'd like to bring up discussion here for `link` and `unlink` before making any further changes. 

IMO `link` and `unlink` are for creating/deleting the through instances i.e no target instances should be involved.
If so, the question is how can we make sure the target exists? Or is it users' responsibility?

Reply:
My take is we should check target instance exists, if not throw error properly.

Reply:
Can we leave `link` and `unlink` out of scope of this pull request and open a follow-up issue to discuss the desired behavior and implement it?